Codership 5.7.37 merge

venkatesh-prasad-v opened the pull request:
Implemented PXC-3942 (Merge codership/mysql-wsrep 5.7.37 with conflicts)

https://jira.percona.com/browse/PXC-3942

Merge with conflicts.

Merge tag 'wsrep_5.7.37-25.29' into codership_5.7.37_merge

Conflicts:
            WSREP_VERSION
            mysql-test/suite/galera/r/galera_toi_drop_database.result
            mysql-test/suite/galera/t/galera_toi_drop_database.test
            mysql-test/suite/galera_3nodes/t/galera_garbd.test
            sql/wsrep_sst.cc

***
 - Merge conflicts resolved.
    
- Updated WSREP_PATCH_VERSION to 59 in WSREP_VERSION file.

- Updated WSREP-REVISION to reflect the upstream refresh revision
 commit id: 8b2028b


Testing Done
----
*Trial 1:*
Jenkins: https://pxc.cd.percona.com/job/pxc-5.7-param/312/console
Failing tests: `galera_3nodes.galera_garbd` [Fixed]

*Trial 2:*
Galera run: https://pxc.cd.percona.com/view/PXC%205.7/job/pxc-5.7-param/321/console
Failing tests: None

Full MTR: https://pxc.cd.percona.com/job/pxc-5.7-param/318/console
Failing tests: In Progress

temeo committed:
MTR fixes to make tests work with binary packages

- Backported garbd executable detection and checks.
- Backported galera_3nodes.galera_garbd.
- Add warning suppression for wsrep.wsrep_mysqld_safe.
  The warning appears when running MTR in AWS.
- Made galera.galera_toi_drop_database deterministic. Replaced big
  autocommit inserts with short transactions which are left open to be
  BF aborted by DROP DATABASE.

Other:
- Cleaned up support-files/wsrep.cnf which is generated file.

sjaakola committed:
codership/mysql-wsrep-bugs#1193 innodb_disallow_writes refactoring

This commit contains a test for checking if disk writes may leak in mysql data directory when node is in SST donor state.
The test uses garbd daeomon to launch SST request with method: 'backup', which is handled by new SST script: 'wsrep_sst_backup'.
This backup script may be later extended as general purpose backup method for the cluster.

Donor thread handler, in wsrep_sst.cc, has new dbug sync point right after the donor state has been reached and disk writes are supposed to be paused.

innodb os0file.cc has also one change where WAIT_ALLOW_WRITES is moved to cover only write requests.
Earlier location would block also reads in dono state.

Fixed race condition happening in wsrep_sst_rsync:
* wsrep_rsync_sst script requests for flush tables,
  and then waits in a loop until mysqld has created file tables_flushed,
  as confirmation that FLUSH TABLES has completed
* mysqld's SST donor thread, wakes for the flush table request and then performs FTWRL,
  and after this it creates the tables_flushed file
* note that SST script will now continue to startup rsync sending
* mysqld's SST donor thread now calls for sst_disallow_writes(),
  so that innodb would setup disk IO blockage, however rsyncing may already be ongoing at this point

This race condition is fixed in this commit, by performing all disk IO blocking before
creating the tables_flushed file.

sjaakola committed:
Merge tag 'mysql-5.7.37' into 5.7-v25-mysql-5.7.37

sjaakola committed:
Fixed garbd log file location

sjaakola committed:
disabled galera_3nodes.galera_garbd_backup test

This test should be enabled, when donor state can effectively stop all disk IO

sjaakola committed:
Configuring wsrep_sst_backup script only for testing usage

temeo committed:
Merged man pages from upstream 5.7.37

temeo committed:
Bump version number to 5.7.37-25.29

Commit:
Implemented PXC-3942 (Merge codership/mysql-wsrep 5.7.37 with conflicts)

https://jira.percona.com/browse/PXC-3942

Merge with conflicts.

Merge tag 'wsrep_5.7.37-25.29' into codership_5.7.37_merge

Conflicts:
        WSREP_VERSION
        mysql-test/suite/galera/r/galera_toi_drop_database.result
        mysql-test/suite/galera/t/galera_toi_drop_database.test
        mysql-test/suite/galera_3nodes/t/galera_garbd.test
        sql/wsrep_sst.cc

Commit:
Implemented PXC-3942 (Merge codership/mysql-wsrep 5.7.37 conflicts resolved)

https://jira.percona.com/browse/PXC-3942

Merge conflicts resolved.

***
Updated WSREP_PATCH_VERSION to 59 in WSREP_VERSION file.

***
Updated WSREP-REVISION to reflect the upstream refresh revision
commit id: 8b2028b

it-percona-cla commented:
[![CLA assistant check](https://cla-assistant.percona.com:443/pull/badge/not_signed)](https://cla-assistant.percona.com:443/percona/percona-xtradb-cluster?pullRequest=1649) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.percona.com:443/percona/percona-xtradb-cluster?pullRequest=1649) before we can accept your contribution.<br/>**1** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: venkatesh-prasad-v<br/>:x: sjaakola<br/>:x: temeo<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.percona.com:443/check/percona/percona-xtradb-cluster?pullRequest=1649) it.</sub>

Commit:
Implemented PXC-3942 (Merge codership/mysql-wsrep 5.7.37, fix MTR failures)

***
Improve the automatic detection of garbd executable and adjusted the
tests that use garbd.

***
The test `galera.galera_garbd` is re-enabled.
